cql3/statements/ks_prop_defs.cc

```diff
@@ -74,20 +75,64 @@ std::optional<sstring> ks_prop_defs::get_replication_strategy_class() const {
     return _strategy_class;
 }
 
-lw_shared_ptr<keyspace_metadata> ks_prop_defs::as_ks_metadata(sstring ks_name) {
+std::map<sstring, sstring> ks_prop_defs::prepare_options(const sstring& strategy_class, const std::map<sstring, sstring>& old_options = {}) {
     auto options = get_replication_options();
     options.erase(REPLICATION_STRATEGY_CLASS_KEY);
-    return keyspace_metadata::new_keyspace(ks_name, get_replication_strategy_class().value(), options, get_boolean(KW_DURABLE_WRITES, true));
+
+    if (strategy_class != "NetworkTopologyStrategy") {
+        return options;
+    }
+
```

Comment:
Seems wrong to add strategy-specific code here. Maybe add a virtual abstract_replication_strategy::adjust_options()?

Reply:
Thanks for the review!

Adding a virtual method to `abstract_replication_strategy` was my initial idea too. Unfortunately, that would probably require a lot of refactoring. Let me explain.

Currently, to create an `abstract_replication_strategy` (abbreviated as `a_r_s` from now on), we need to pass the replication strategy options.
1. each time `a_r_s` is instantiated in our code, options are passed down from some `keyspace_metadata` instance.
2. `keyspace_metadata` is created in many places. The options used to instantiate a `keyspace_metadata` can come from:
- a user performing a CQL query, in `ks_prop_defs::as_ks_metadata` and `ks_prop_defs::as_ks_metadata_update`, in cql3/statements/ks_prop_defs.cc
- a user accessing the DB through Thrift, in `thrift_handler::keyspace_from_thrift`, in thrift/handler.cc
- some old keyspace_metadata, e.g. `database::update_keyspace` in database.cc
- a developer, e.g. `service::create_keyspace_if_missing`, in auth/service.cc
3. there is only one case when the options should be expanded/adjusted: the CQL case. This is when Cassandra does it. Remember, expanding the replication_factor is a UX/UI convenience thing for the CQL API.
4. since `a_r_s` is sometimes created with the options expanded and sometimes without, the decision to adjust the options must be made before `a_r_s` is created.
5. therefore adjusting the options can't be made in a `a_r_s` (non-static) method, as it requires an `a_r_s` instance, which we cannot yet create. QED.

I could try to refactor the code somehow so that a virtual method in `a_r_s` could be used, or come up with a different "generic"/"modular" solution. But is it worth it? To answer this question, I looked into how Cassandra does it:
https://github.com/apache/cassandra/commit/12aa45434a6d5913a2a3db91a26a8d2f7081042a
it obtains a `Class<? extends AbstractReplicationStrategy>` from the given strategy string and then uses reflection to check if it has a `prepareOptions` static method.
Great, so it's a nice generic solution: every ReplicationStrategy can have its own `prepareOptions` static method; if it doesn't then the options stay unchanged. However, the patch was made a year ago, and the only class which has a `prepareOptions` method to this day is, of course, `NetworkTopologyStrategy`, which does the replication_factor expansion. For me it's a great example of over-engineering, or [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it). I prefer the simple approach.

Reply:
We could replace the convoluted way we create the replication strategy (which in turn was derived from Java) by registering a factory class with two virtual methods, one to massage the parameters and one to create the class. But I agree it's overkill.